lib/src/optimism/batcher_db.rs

```diff
@@ -0,0 +1,149 @@
+}
+
+impl MemDb {
+    pub fn new() -> Self {
+        MemDb {
+            full_op_block: HashMap::new(),
+            op_block_header: HashMap::new(),
+            full_eth_block: HashMap::new(),
+            eth_block_header: HashMap::new(),
+        }
+    }
+}
+
+impl Default for MemDb {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl BatcherDb for MemDb {
+    fn get_full_op_block(&mut self, block_no: u64) -> Result<BlockInput<OptimismTxEssence>> {
+        let op_block = self.full_op_block.remove(&block_no).unwrap();
+        assert_eq!(block_no, op_block.block_header.number);
+
+        // Validate tx list
+        {
+            let mut tx_trie = MptNode::default();
+            for (tx_no, tx) in op_block.transactions.iter().enumerate() {
+                let trie_key = tx_no.to_rlp();
+                tx_trie.insert_rlp(&trie_key, tx)?;
+            }
+            if tx_trie.hash() != op_block.block_header.transactions_root {
+                bail!("Invalid op block transaction data!")
+            }
+        }
+
+        Ok(op_block)
+    }
+
+    fn get_op_block_header(&mut self, block_no: u64) -> Result<Header> {
+        let op_block = self.op_block_header.remove(&block_no).unwrap();
+        assert_eq!(block_no, op_block.number);
+
+        Ok(op_block)
+    }
+
+    fn get_full_eth_block(&mut self, block_no: u64) -> Result<BlockInput<EthereumTxEssence>> {
+        let eth_block = self.full_eth_block.remove(&block_no).unwrap();
+        assert_eq!(block_no, eth_block.block_header.number);
+
+        // Validate tx list
+        {
+            let mut tx_trie = MptNode::default();
+            for (tx_no, tx) in eth_block.transactions.iter().enumerate() {
+                let trie_key = tx_no.to_rlp();
+                tx_trie.insert_rlp(&trie_key, tx)?;
+            }
+            if tx_trie.hash() != eth_block.block_header.transactions_root {
+                bail!("Invalid eth block transaction data!")
+            }
+        }
+
+        // Validate receipts
+        if eth_block.receipts.is_some() {
+            let mut receipt_trie = MptNode::default();
+            for (tx_no, receipt) in eth_block.receipts.as_ref().unwrap().iter().enumerate() {
+                let trie_key = tx_no.to_rlp();
+                receipt_trie.insert_rlp(&trie_key, receipt)?;
+            }
+            if receipt_trie.hash() != eth_block.block_header.receipts_root {
+                bail!("Invalid eth block receipt data!")
+            }
+        } else {
+            let can_contain_deposits = deposits::can_contain(
+                &OPTIMISM_CHAIN_SPEC.deposit_contract,
+                &eth_block.block_header.logs_bloom,
+            );
+            let can_contain_config = system_config::can_contain(
+                &OPTIMISM_CHAIN_SPEC.system_config_contract,
+                &eth_block.block_header.logs_bloom,
+            );
+            assert!(!can_contain_deposits);
+            assert!(!can_contain_config);
```

Comment:
Is this an actual (debug) assert? then we should remove it or only run it when not on the zkvm. If this an validity error, we should bail instead.

Reply:
It is a validity error; you are right, it should bail instead of panic. Done